### PR TITLE
[CU-30xvv59] Added new Out-Given-In for CPP

### PR DESCRIPTION
**Pull request by PoisonPhang**
## Issue
[[CU-30xvv59]](https://app.clickup.com/t/30xvv59)

## Description
Note: This function is not in use yet

### CODE REVIEW

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-30xvv59 5.2) 2. Out-given-in with Fee (eq: 2 + Algorithm: constant_product::compute_out_given_in)](https://app.clickup.com/t/30xvv59)

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs [hunk 1]
```diff
@@ -86,6 +86,59 @@ where
...
+	let value = b_i.safe_add(&a_sent.safe_mul(&left_from_fee)?)?;
+	let value = b_i.safe_div(&value)?;
+	let value = value.checked_powd(weight_ratio).ok_or(ArithmeticError::Overflow)?;
+	let value = Decimal::ONE.safe_sub(&value)?;
+
+	let a_out = b_o.safe_mul(&value)?.to_u128().ok_or(ArithmeticError::Overflow)?;
+	let fee = a_sent
+		.safe_mul(&Decimal::ONE.safe_sub(&left_from_fee)?)?
+		.to_u128()
+		.ok_or(ArithmeticError::Overflow)?;
```
Comment (hussein-aitlahcen) [inline code/parachain/frame/composable-maths/src/dex/constant_product.rs:137]: you can introduce an intermediate `let a_sent_fee_cut = a_sent.safe_mul(&left_from_fee)` to then just do a `let fee = a_sent.safe_sub(a_sent_fee_cut)` to get the fee with the delta instead of recomputing using mul

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs [hunk 2]
```diff
@@ -86,6 +86,59 @@ where
 	Ok(ao)
 }
 
+/// Compute the amount of the output token given the amount of the input token.
+///
+/// Returns a tuple containing `(a_out, fee)`.
+/// To get `a_out` without accounting for the fee, set `f = 0`.
+///
+/// **NOTE:** Weights must be normalized.
```
Comment (vimukthi-git) [inline code/parachain/frame/composable-maths/src/dex/constant_product.rs:94]:
```suggestion
/// **NOTE:** Weights must be already normalized.
```

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs [hunk 3]
```diff
@@ -86,6 +86,59 @@ where
...
+///
+/// From https://balancer.fi/whitepaper.pdf, equation (15)
```
Comment (vimukthi-git) [inline code/parachain/frame/composable-maths/src/dex/constant_product.rs:96]: Let's refer our own doc as that has the fee included (while also having this reference) and we keep it up to date

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs [hunk 4]
```diff
@@ -86,6 +86,59 @@ where
...
+///
+/// # Paramaters
+/// * `w_i` - Weight of the input token
+/// * `w_o` - Weight of the output token
+/// * `b_i` - Balance of the input token
+/// * `b_o` - Balance of the output token
+/// * `a_sent` - Amount of the input token sent by the user
+/// * `f` - Total swap fee
+pub fn compute_out_given_in_new<T: PerThing>(
```
Comment (vimukthi-git) [inline code/parachain/frame/composable-maths/src/dex/constant_product.rs:105]: Need unit tests

Comment (PoisonPhang) [reply code/parachain/frame/composable-maths/src/dex/constant_product.rs:106]: bf3e802daebf5e1a7f5a8b6f200232e407955e78
Also going to add a test for divide by zero check

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs (hunk 3 above)
Comment (PoisonPhang) [reply code/parachain/frame/composable-maths/src/dex/constant_product.rs]: c932d03d98fc89a809c7d0c6fa8839067e2a06f8

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs (hunk 2 above)
Comment (PoisonPhang) [reply code/parachain/frame/composable-maths/src/dex/constant_product.rs]: c932d03d98fc89a809c7d0c6fa8839067e2a06f8

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs (hunk 4 above)
Comment (PoisonPhang) [reply code/parachain/frame/composable-maths/src/dex/constant_product.rs:106]: 922cdbdf021ee84d8adeb3c02a61c69ed0627de8

File: code/parachain/frame/composable-maths/src/dex/tests.rs [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+		use super::*;
+
+		/// should always be equal to the number of elements in the `list` array found in
+		/// `checked_inputs_and_outputs` function
+		const CHECKED_I_AND_O_SIZE: u32 = 5;
+
+		prop_compose! {
+			/// Returns (w_i, w_o, b_i, b_o, a_sent, fee, a_out, fee_out)
+			fn checked_inputs_and_outputs()
+			(x in 0..CHECKED_I_AND_O_SIZE) -> (Permill, Permill, u128, u128, u128, Permill, u128, u128) {
```
Comment (vimukthi-git) [inline code/parachain/frame/composable-maths/src/dex/tests.rs:19]: Another nice proptest would be to check on a generated range of possible values for weights, balances and input and see if they cause unexpected errors.

**Review by vimukthi-git — APPROVED** [review]
LGTM

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs [hunk 6]
```diff
@@ -86,6 +86,58 @@ where
...
+///
+/// # Parameters
+/// * `w_i` - Weight of the input token
+/// * `w_o` - Weight of the output token
+/// * `b_i` - Balance of the input token
+/// * `b_o` - Balance of the output token
+/// * `a_sent` - Amount of the input token sent by the user
+/// * `f` - Total swap fee
+///
+/// From https://github.com/ComposableFi/composable/blob/cu-2yyx1w9/rfcs/0008-pablo-lbp-cpp-restructure.md#41-fee-math-updates,
```
Comment (vimukthi-git) [inline code/parachain/frame/composable-maths/src/dex/constant_product.rs:104]:
```suggestion
/// From https://github.com/ComposableFi/composable/blob/main/rfcs/0008-pablo-lbp-cpp-restructure.md#41-fee-math-updates,
```

File: code/parachain/frame/composable-maths/src/dex/tests.rs (hunk 5 above)
Comment (PoisonPhang) [reply code/parachain/frame/composable-maths/src/dex/tests.rs:19]: While I agree it would be nice to test this on a larger dataset, random or pseudo-random numbers may lead to a flaky test.

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs (hunk 6 above)
Comment (PoisonPhang) [reply code/parachain/frame/composable-maths/src/dex/constant_product.rs]: a5b0ad28f92311f83ca64cd5568cbbb412e52f1e

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs [hunk 7]
```diff
@@ -86,6 +86,58 @@ where
...
+/// * `w_i` - Weight of the input token
+/// * `w_o` - Weight of the output token
+/// * `b_i` - Balance of the input token
+/// * `b_o` - Balance of the output token
+/// * `a_sent` - Amount of the input token sent by the user
+/// * `f` - Total swap fee
+///
+/// From https://github.com/ComposableFi/composable/blob/main/rfcs/0008-pablo-lbp-cpp-restructure.md#41-fee-math-updates,
+/// equation (2)
+pub fn compute_out_given_in_new<T: PerThing>(
```
Comment (PoisonPhang) [inline code/parachain/frame/composable-maths/src/dex/constant_product.rs:106]: Should round down overall for OutGivenIn
https://github.com/balancer-labs/balancer-v2-monorepo/blob/592b0534f782b9bfb668990f7821e041a9b775d4/pkg/pool-weighted/contracts/WeightedMath.sol#L95

File: code/parachain/frame/composable-maths/src/dex/tests.rs (hunk 5 above)
Comment (vimukthi-git) [reply code/parachain/frame/composable-maths/src/dex/tests.rs:19]: does not have to be a random generator, just need some specific ranges to be tested with the combinations possible. But not that critical. Feel free to resolve.

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs (hunk 7 above)
Comment (PoisonPhang) [reply code/parachain/frame/composable-maths/src/dex/constant_product.rs:107]: 7b5b4cabd92cfac17dc1c0b66b5e4cd85e92dce7

File: code/parachain/frame/composable-maths/src/dex/tests.rs (hunk 5 above)
Comment (PoisonPhang) [reply code/parachain/frame/composable-maths/src/dex/tests.rs]: f62b0bf1ae1ba0a971de22aea3dbe0c34a82db76

**Review by vimukthi-git — APPROVED** [review]

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs [hunk 8]
```diff
@@ -91,6 +91,88 @@ where
...
+	} else {
+		Decimal::from(f.left_from_one().deconstruct().into())
+			.safe_div(&Decimal::from(T::one().deconstruct().into()))?
+	};
+	let a_sent_fee_cut = a_sent.safe_mul(&left_from_fee)?;
+
+	let value = b_i.safe_add(&a_sent_fee_cut)?;
+	let value = b_i.safe_div(&value)?;
+	let value = value.checked_powd(weight_ratio).ok_or(ArithmeticError::Overflow)?;
+	let value = Decimal::ONE.safe_sub(&value)?;
```
Comment (benluelo) [inline code/parachain/frame/composable-maths/src/dex/constant_product.rs:139]: Same comment as the other PR - rename `value` here? This also seems to be a common-ish pattern, perhaps it could be extracted out somehow?

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs [hunk 9]
```diff
@@ -91,6 +91,88 @@ where
...
+	w_o: T,
+	b_i: u128,
+	b_o: u128,
+	a_sent: u128,
+	f: T,
+) -> Result<(u128, u128), ArithmeticError> {
+	let w_i = Decimal::from(w_i.deconstruct().into());
+	let w_o = Decimal::from(w_o.deconstruct().into());
+	let b_i = Decimal::from(b_i);
+	let b_o = Decimal::from(b_o);
```
Comment (benluelo) [inline code/parachain/frame/composable-maths/src/dex/constant_product.rs:123]: use `from_u128` here

File: code/parachain/frame/composable-maths/src/dex/tests.rs [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+
+				CHECKED_I_AND_O_LIST[x]
+			}
+		}
+
+		prop_compose! {
+			#[allow(clippy::useless_conversion)]
+			fn range_inputs()
+			(
+				w_i in Range::<u32>::from(1..100),
```
Comment (benluelo) [inline code/parachain/frame/composable-maths/src/dex/tests.rs:93]:
```suggestion
				w_i in 1..100_u32,
```

Comment (benluelo) [reply code/parachain/frame/composable-maths/src/dex/tests.rs:93]: I think this should work for type inference

File: code/parachain/frame/composable-maths/src/dex/tests.rs [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+			-> InputsAndOutputs {
+				InputsAndOutputs {
+					w_i: Permill::from_percent(w_i),
+					w_o: Permill::from_percent(w_o),
+					b_i,
+					b_o,
+					a_sent,
+					f: Permill::from_parts(f),
+					a_out: u128::default(),
+					fee: u128::default(),
```
Comment (benluelo) [inline code/parachain/frame/composable-maths/src/dex/tests.rs:113]: using `0` here might be clearer

File: code/parachain/frame/composable-maths/src/dex/constant_product.rs (hunk 8 above)
Comment (PoisonPhang) [reply code/parachain/frame/composable-maths/src/dex/constant_product.rs:139]: Will go ahead and rename them, we can extract the functionality to a single function once all the code is present in the main. Otherwise, the refactoring and merge conflicts will be a pain.

**Review by benluelo — APPROVED** [review]
